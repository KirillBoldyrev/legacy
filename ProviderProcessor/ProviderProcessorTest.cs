using ApprovalTests;
using ApprovalTests.Reporters;
using FakeItEasy;
using Newtonsoft.Json;
using NUnit.Framework;
using ProviderProcessing.ProcessReports;
using ProviderProcessing.ProviderDatas;
using ProviderProcessing.References;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ProviderProcessor
{
    [TestFixture]
    public class ProviderProcessorTest
    {
        private ProviderProcessing.ProviderProcessor providerProcessor { get; set; }

        [SetUp]
        public void SetUp()
        {
            var productsReference = A.Fake<ProductsReference>();
            A.CallTo(() => productsReference.FindCodeByName("Product111")).Returns(111);
            A.CallTo(() => productsReference.FindCodeByName("Product222")).Returns(222);
            A.CallTo(() => productsReference.FindCodeByName("Product333")).Returns(null);

            var measureUnitsReference = A.Fake<MeasureUnitsReference>();
            A.CallTo(() => measureUnitsReference.FindByCode("measure01")).Returns(new MeasureUnit());
            A.CallTo(() => measureUnitsReference.FindByCode("measure02")).Returns(null);

            providerProcessor = A.Fake<ProviderProcessing.ProviderProcessor>();
            A.CallTo(() => providerProcessor.GetProductsReferenceInstance()).Returns(productsReference);
            A.CallTo(() => providerProcessor.GetMeasureUnitsReferenceInstance()).Returns(measureUnitsReference);
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void Validation_Test_PricesAndMeasures()
        {
            var prices = new decimal[] { -100, 0, 100 };
            var measureUnits = new string[] { "measure01", "measure02" };
            var products = new List<ProductData>();

            foreach (var p in prices)
            {
                foreach (var m in measureUnits)
                {
                    products.Add(new ProductData() { Price = p, MeasureUnitCode = m });
                }
            }

            var validationResult = new List<ProductValidationResult>();

            foreach (var p in products)
            {
                var vResult = providerProcessor.ValidatePricesAndMeasureUnitCodes(p);

                validationResult.AddRange(vResult);
            }

            Approvals.Verify(JsonConvert.SerializeObject(validationResult));
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void Validation_Test_Names()
        {
            var productNames = new string[] { "Product111", "Product222", "Product333" };
            var products = new List<ProductData>();

            foreach (var pName in productNames)
            {
                products.Add(new ProductData() { Name = pName });
            }

            var validationResult = new List<ProductValidationResult>();

            foreach (var p in products)
            {
                var vResult = providerProcessor.ValidatePricesAndMeasureUnitCodes(p);

                validationResult.AddRange(vResult);
            }

            Approvals.Verify(JsonConvert.SerializeObject(validationResult));
        }
    }
}
