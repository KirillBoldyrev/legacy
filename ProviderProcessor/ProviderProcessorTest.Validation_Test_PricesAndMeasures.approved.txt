[{"Product":{"Id":"00000000-0000-0000-0000-000000000000","Name":null,"Price":-100.0,"MeasureUnitCode":"measure01"},"Message":"Bad price","Severity":1},{"Product":{"Id":"00000000-0000-0000-0000-000000000000","Name":null,"Price":-100.0,"MeasureUnitCode":"measure02"},"Message":"Bad price","Severity":1},{"Product":{"Id":"00000000-0000-0000-0000-000000000000","Name":null,"Price":-100.0,"MeasureUnitCode":"measure02"},"Message":"Bad units of measure","Severity":1},{"Product":{"Id":"00000000-0000-0000-0000-000000000000","Name":null,"Price":0.0,"MeasureUnitCode":"measure01"},"Message":"Bad price","Severity":1},{"Product":{"Id":"00000000-0000-0000-0000-000000000000","Name":null,"Price":0.0,"MeasureUnitCode":"measure02"},"Message":"Bad price","Severity":1},{"Product":{"Id":"00000000-0000-0000-0000-000000000000","Name":null,"Price":0.0,"MeasureUnitCode":"measure02"},"Message":"Bad units of measure","Severity":1},{"Product":{"Id":"00000000-0000-0000-0000-000000000000","Name":null,"Price":100.0,"MeasureUnitCode":"measure02"},"Message":"Bad units of measure","Severity":1}]